Add Operator.from_circuit constructor method

mtreinish opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

This commit adds a new constructor method .from_circuit() to the
Operator class. This allows users to construct Operator objects from
QuantumCircuits and have additional features than the default __init__
constructor allows. To start this just involves permuting the qubit
order based on the specified layout. The permutation based on layout has
come before, especially in testing with transpilation (see #7213 for
where equivalent functionality was done inline in the unittests).
Exposing this as an api will provide a single place to build an Operator
with a layout. It also will provide us a dedicated location if we need
to offer additional options to the construction of Operators from
circuits.

### Details and comments

jakelishman commented on qiskit/quantum_info/operators/operator.py:
```diff
@@ -199,6 +199,49 @@ def from_label(cls, label):
+    def from_circuit(cls, circuit, use_layout=True, layout=None):
+        """Create a new Operator object from a :class`.QuantumCircuit`
+
+        While a :class:`.QuantumCircuit` object can passed directly as ``data``
+        to the class constructor
```
I think a clause has gone missing here.

mtreinish replied:
Fixed in: https://github.com/Qiskit/qiskit-terra/pull/7616/commits/b6ebe91668be88faf76dca608a623d46a1770135

jakelishman commented on qiskit/quantum_info/operators/operator.py:
```diff
@@ -199,6 +199,49 @@ def from_label(cls, label):
+        Args:
+            circuit (QuantumCircuit): The :class:`.QuantumCircuit` to create an Operator
+                object from.
+            use_layout (bool): When set to ``False`` if the input ``circuit``
+                has a layout set it will be ignored
+            layout (Layout): If specified this kwarg can be used to specify a
+                particular layout to use to permute the qubits in the created
+                :class:`.Operator`. If this is specified it will be used instead
+                of a layout contained in the ``circuit`` input.
```
It's slightly unusual to me that `use_layout` is ignored if `layout` is passed, given how similar they are.  I'm not at all married to the idea, but how would it feel to make the type of `layout` a Boolean or a `Layout` - `True` means take from the circuit if available, `False` means ignore all, and `Layout` means use this as an override?

mtreinish replied:
I can change it if you'd prefer that pattern. TBH, that seems a bit weirder to have mixed types on a single arg, but there really isn't a good pattern here. Maybe rename `use_layout` to something like `use_set_layout` or `use_bundled_layout` to indicate it means use the `_layout` property that gets set in the circuit by `transpile()`. Or alternatively we can just remove it as a flag because using the specified layout set layout is really the only difference between `Operator(circuit)` and `Operator.from_circuit(circuit)` so instead of having a flag we can just tell people to call `__init__` if that's the behavior they want.

jakelishman replied:
No, I think your method is better.  Perhaps flip it, and call it `ignore_set_layout`?  That's maybe cleaner - it's slightly odd that `Operator.from_circuit(qc, layout=..., use_layout=True)` will ignore `use_layout`, but it seems more natural that `Operation.from_circuit(qc, layout=..., ignore_set_layout=True)` won't matter.

mtreinish replied:
Done in: https://github.com/Qiskit/qiskit-terra/pull/7616/commits/b6ebe91668be88faf76dca608a623d46a1770135

jakelishman commented on releasenotes/notes/Operator-from_circuit-25b20d4b3ad5c398.yaml:
```diff
@@ -0,0 +1,28 @@
+    :meth:`.Operator.from_circuit` for creating a new :class`.Operator`
+    object from a :class:`.QuantumCircuit`. While this was possible normally
+    using the default constructor, the :meth:`.Operator.from_circuit` method
+    provide additional options to adjust how the operator is created. Primarily
```
```suggestion
    provides additional options to adjust how the operator is created. Primarily
```

jakelishman reviewed (commented):
Looks like a very nice new feature!  I had one minor question about interface, but I'm not 100% convinced it's really better.

The tests seem nice and thorough - the only additional one I think might be missing is one if `use_layout=False` but the circuit has a layout set.

mtreinish commented:
> The tests seem nice and thorough - the only additional one I think might be missing is one if `use_layout=False` but the circuit has a layout set.

I added a test for this (well for `ignore_set_layout` which is what we changed `use_layout` to) in: https://github.com/Qiskit/qiskit-terra/pull/7616/commits/b6ebe91668be88faf76dca608a623d46a1770135

ikkoham commented:
Do you think should we deprecate `__init__` from QuantumCircuit? I think two ways are duplicated.

jakelishman commented:
Sorry Ikko, I should have left this open for a bit longer to give you time to comment.

I think there's probably lots of places that use `Operator(qc)` in user code already, to the degree that deprecating it and forcing people to use a more verbose interface when they don't need to is probably quite inconvenient without too much benefit.  If I'd thought about it a bit more when reviewing, though, it would have been good to de-duplicate the code between `_init_instruction` and `from_circuit` a little more.

At any rate, we wouldn't be able to deprecate the `__init__` path until Terra 0.21 at the very least - the alternative needs to be in place for at least one minor version before we start issuing warnings.

ikkoham commented:
Thank you very much. I agree with your opinion. For now, let's leave it as it is until something goes wrong.
